File: Leetcode/543-Diameter-Of-Binary-Tree/DiameterOfBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int helper(TreeNode* root, int & dia) {
        if (root == NULL)
            return (0);

        int l = helper(root->left, dia);
        int r = helper(root->right, dia);

        dia = max(dia, l + r);

        return (1 + max(l, r));

    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        int dia = 0;
        (void) helper(root, dia);
        return (dia);
    }
};

File: Leetcode/543-Diameter-Of-Binary-Tree/DiameterOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    var getDiameter = function(node) {
        var res = [0,0]
        if (!node) {
            return res
        }
        var left = getDiameter(node.left)
        var right = getDiameter(node.right)
        res[0] = Math.max(left[1]+right[1], Math.max(left[0], right[0]))
        res[1] = Math.max(left[1], right[1])+1
        return res
    };
    var res = getDiameter(root)
    return res[0]
};
